tests: add auth tests

edsantiago opened the pull request:
Add authentication tests using local registry:
 - add output string checking wherever possible; this
   caught a bug in the old tests, a command that was
   failing due to a typo instead of the expected way.

 - new --tls-verify=true tests, if $BUILDAH_AUTHDIR is
   defined by caller. This will require infrastructure
   changes in many places: CI, Fedora gating tests,
   and RHEL gating tests. I choose to defer that
   work for a future PR. (Tests are in this PR,
   and pass on manual run, but will be skipped
   until said changes are made)

 - new buildah-login tests, positive and negative

Also:
 - remove unused tests in from.bats

 - remove unused and unusable and misleading .sh file (probably
   a holdover from early docker days)

 - add "authenticate: " prefix to all tests in that file,
   to make it easier for viewer to find the test source file

Signed-off-by: Ed Santiago <santiago@redhat.com>

/kind cleanup

edsantiago commented on tests/authenticate.bats:
```diff
@@ -2,41 +2,48 @@
-  # This should fail
-  run_buildah 125 push  --signature-policy ${TESTSDIR}/policy.json --tls-verify=true localhost:5000/my-alpine
+  # With tls-verify=true, should fail due to self-signed cert
+  run_buildah 125 push  --signature-policy ${TESTSDIR}/policy.json --tls-verify=true alpine localhost:5000/my-alpine
```
This is why we check the output of our tests. As previously written, the test was failing with exit status 125 (as expected), but the error was:
```
error pushing image "localhost:5000/my-alpine" to "docker://localhost:5000/my-alpine": error locating image with name "localhost:5000/my-alpine" ([localhost:5000/my-alpine]): image not known
```

TomSweeneyRedHat replied:
I think this was a veiled slap of @TomSweeneyRedHat 's hands....   Well deserved I might add!

edsantiago replied:
Eek - that was never my intent. I didn't look at `git blame` nor did I mean to single anyone out. This is just a common pattern I see in all the tests.

edsantiago commented:
AFAICT tests are failing because of this code in `contrib/cirrus/lib.sh`:
```bash
    certdirpath=/etc/docker/certs.d
    ...
    cp $authdirpath/domain.crt $certdirpath/localhost:5000/ca.crt
    cp $authdirpath/domain.crt $certdirpath/localhost:5000/domain.crt
```
...in combination with the (apparent?) fact that `buildah`, which documents a default `--cert-dir` of `/etc/containers/certs.d`, is also silently accepting `/etc/docker/certs.d`.

@TomSweeneyRedHat AFAICT the origin of that code for copying certs into `/etc` is your c77a8d39f1c1f9eb751fec18dc04df29865850bc from  December 2017. Would you happen to remember why you copied those certs into `/etc`? Do you (or anyone else) have an objection to removing that?

TomSweeneyRedHat commented:
@edsantiago couple of things.  First off, LGTM over all.
1)  Please do NOT remove the *auth*.sh file that you're removing.  I actually use that from time to time.  If you want to comment it up, or move it to a different directory ("ByHandUnsupportedTests" or some such), I'm fine with that.  We also have other *.sh files in there that I use from time to time.  All that said, the script probably could use some love, I've not used it for a few months.

2) We've podman, skopeo, and buildah login/logout these days.  I'd really love it if we could build a test somewhere that could test all of those at once.  Tall task I know.  Better yet, I'd like to get the login/logout functionality moved into containers/common and then tested from  there.  Nothing for today, but a seed for a thought perhaps.

3) As far as the /etc certs go, I'd all kinds of trouble getting that working.  I don't recall why I put them there specifically, but it was probably the way I got it to work.  I'm fine with removing it, especially if you can stream line it and make it more "real-life".

Thanks for vamping this up @edsantiago , be sure to add a jira card for this....

edsantiago commented:
I'm assuming this is a flake; could someone with privs hit Rerun?
```
         # $ /var/tmp/go/src/github.com/containers/buildah/tests/./../bin/buildah from --quiet --pull=false --signature-policy /var/tmp/go/src/github.com/containers/buildah/tests/./policy.json fedora-minimal
         # The following failures happened while trying to pull image specified by "fedora-minimal" based on search registries in /var/tmp/go/src/github.com/containers/buildah/tests/./registries.conf:
         # * "localhost/fedora-minimal": Error initializing source docker://localhost/fedora-minimal:latest: error pinging docker registry localhost: Get https://localhost/v2/: dial tcp 127.0.0.1:443: connect: connection refused
         # * "docker.io/library/fedora-minimal": Error initializing source docker://fedora-minimal:latest: Error reading manifest latest in docker.io/library/fedora-minimal: errors:
         # denied: requested access to the resource is denied
         # unauthorized: authentication required
         # 
         # * "quay.io/fedora-minimal": Error initializing source docker://quay.io/fedora-minimal:latest: Error reading manifest latest in quay.io/fedora-minimal: error parsing HTTP 404 response body: invalid character '<' looking for beginning of value: "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\n<title>404 Not Found</title>\n<h1>Not Found</h1>\n<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>\n"
         # * "registry.fedoraproject.org/fedora-minimal": Error initializing source docker://registry.fedoraproject.org/fedora-minimal:latest: Error reading manifest latest in registry.fedoraproject.org/fedora-minimal: manifest unknown: manifest unknown
```

rhatdan commented:
bors r+

TomSweeneyRedHat commented:
LGTM, thanks for fixing this up @edsantiago

edsantiago commented:
@TomSweeneyRedHat belated response:

> Please do NOT remove the auth.sh file

Restored. Sorry, it did not look like anything useful (e.g. "This should fail" in comments but not actually in the code).

> [test] podman, skopeo, and buildah login/logout

One of the podman system tests cross-checks against skopeo already; I think this (`authenticate.bats`) would probably be a good place to add a podman/buildah cross-check, since buildah requires podman. I've added a tickler for me to do that one of these days, thank you.

> /etc/certs

Thanks. I removed it, and tweaked a couple of checks that were relying on it (but did not actually really need it).

rhatdan commented:
bors retry

edsantiago commented:
Do I need to rebase on top of #2432 ?

rhatdan commented:
Yes.

edsantiago commented:
buildah is panicking, and I don't think it's anything I did:
```
not ok 137 bud with copy-from referencing the base image
$ buildah bud --no-cache [...] --jobs 4 -t busybox-mt-derived -f tests/bud/copy-from/Dockerfile4 tests/./bud/copy-from
...
         # fatal error: concurrent map writes
         # fatal error: concurrent map writes
         # 
         # goroutine 35 [running]:
         # runtime.throw(0x1147942, 0x15)
         # 	/usr/lib/golang/src/runtime/panic.go:774 +0x72 fp=0xc0003bfcb0 sp=0xc0003bfc80 pc=0x4373d2
         # runtime.mapassign_faststr(0xfbf940, 0xc0004f3440, 0xc000112d41, 0x5, 0x1ad2648)
         # 	/usr/lib/golang/src/runtime/map_faststr.go:291 +0x3fe fp=0xc0003bfd18 sp=0xc0003bfcb0 pc=0x41b2ce
         # github.com/containers/buildah/imagebuildah.(*Executor).startStage(0xc0002d4300, 0xc000367d10, 0x2, 0x7ffd81864a8f, 0x12, 0x0)
         # 	/var/tmp/go/src/github.com/containers/buildah/imagebuildah/executor.go:262 +0x118 fp=0xc0003bfd78 sp=0xc0003bfd18 pc=0xe54898
         # github.com/containers/buildah/imagebuildah.(*Executor).buildStage(0xc0002d4300, 0x13155a0, 0xc0000b8068, 0xc000367ce0, 0xc000332140, 0x2, 0x2, 0x1, 0x0, 0x0, ...)
...
```

rhatdan commented:
@edsantiago  Could you rebase and repush.

edsantiago commented:
All three CI jobs failed with:
```
Error reading manifest latest in registry.fedoraproject.org/fedora-minimal: manifest unknown: manifest unknown
```
I have to assume that it's a flake. Re-pushed to restart tests.

edsantiago commented:
omg CI passed

rhatdan commented:
bors r+
